### DISCUSSION TITLE
How do you manage enviroment variables, and secrets in fastAPI

### DISCUSSION THREAD

**Discussion by vjanz** [fastapi/fastapi#7719] [Questions] [answered]
Just wanted to get any good practice how you deal with enviroment variables, and secrets like database informations(or anything else that is considered as secret).
I am new to this, and I want to get some advice from developers with experience on this case.
And please if you give me some help, that could be great(if possible) with a small explaination(source code).

Thank to this awesome community in advance.

**Comment by chbndrhnns** [comment]
I'd like to point you in a couple of directions:

- I would start with https://pydantic-docs.helpmanual.io/usage/settings.
- Netflix recently released [dispatch](https://github.com/Netflix/dispatch) which is based on FastAPI, [settings example](https://github.com/Netflix/dispatch/blob/develop/src/dispatch/config.py)
- official project template has a [pull request](https://github.com/tiangolo/full-stack-fastapi-postgresql/pull/87) to use pydantic's `BaseSettings` class

Is this a good start for you?

**Comment by jlugao** [comment]
Also, take a look at [dynaconf ](https://github.com/rochacbruno/dynaconf) and [decouple](https://github.com/henriquebastos/python-decouple)

**Comment by yeyeric** [comment]
took me a while to understand how to manage that properly, here is my way of managing them:
- create a .env file in a /settings folder (git ignore this file), where I put all my variables stored in a format my_variable = "value"
- use pydantic as said above which relies on python-dotenv, specifying my .env file and a class "Settings" with all environment variables which I can use everywhere needed by importing settings 
- in my cicd (i'm using gitlab), I create my environment variables with an "APP_" prefix for all the variables which I'd like to have accessible in my app, and I use the scope from gitlab to determine whether the values are for staging or production branch
- during my cicd, I generate a .env file which exports all my env variables stored within gitlab and available on my project and then keep only the one with "APP_" prefix, removing the prefix. This gives this (I need also to replace "declare -x " because it's added during the export due to the image i'm using, so it may depend on your case):
`export | grep APP_ | sed -e 's/APP_//g' | sed -e 's/declare -x //g' > ./app/settings/.env`

and voilà, nothing complicated for me

P.S. I'm not Facebook nor Google so it might be not top of the top but for my needs this is very easy to manage and secure enough.

Hope it helps

**Comment by tiangolo** [comment, marked as answer]
Thanks for the help here everyone! :cake: 

Thanks @vjanz for closing the issue.

You might also want to check the new docs for settings and env vars: https://fastapi.tiangolo.com/advanced/settings/
